### PR TITLE
feat: Use id from https://graph.microsoft.com/v1.0/me as oidc subject.

### PR DESCRIPTION
**Pull request by splaunov**
Use id from https://graph.microsoft.com/v1.0/me as a subject identifier instead of userinfo sub field.

## Related issue(s)
#2150

## Checklist

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [x] I have referenced an issue containing the design document if my change
      introduces a new feature.
- [x] I am following the
      [contributing code guidelines](../blob/master/CONTRIBUTING.md#contributing-code).
- [x] I have read the [security policy](../security/policy).
- [x] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security. vulnerability, I
      confirm that I got green light (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [ ] I have added tests that prove my fix is effective or that my feature
      works.
- [ ] I have added or changed [the documentation](docs/docs).

### CODE REVIEW

**Comment by splaunov** [comment]
Please give your feedback.
Not tested this changes yet.

File: selfservice/strategy/oidc/provider_microsoft.go [hunk 1]
```diff
@@ -67,7 +69,49 @@ func (m *ProviderMicrosoft) Claims(ctx context.Context, exchange *oauth2.Token)
...
+		defer resp.Body.Close()
+
+		var user struct {
+			Id string `json:"id"`
+		}
+		if err := json.NewDecoder(resp.Body).Decode(&user); err != nil {
+			return nil, errors.WithStack(herodot.ErrInternalServerError.WithReasonf("%s", err))
+		}
+
+		claims.Subject = user.Id
```
Comment (aeneasr) [inline selfservice/strategy/oidc/provider_microsoft.go:111]: Is the value from the `me` endpoint different than the one from the userinfo endpoint?

**Review by aeneasr — COMMENTED** [review]
Thank you! I've got a question

File: selfservice/strategy/oidc/provider_microsoft.go (hunk 1 above)
Comment (splaunov) [reply selfservice/strategy/oidc/provider_microsoft.go:111]: Yes, they are different. Have checked this with my own account.

Comment (aeneasr) [reply selfservice/strategy/oidc/provider_microsoft.go:111]: Could you share what their formats look like? And is there no way to get this other ID from the userinfo endpoint?

Comment (splaunov) [reply selfservice/strategy/oidc/provider_microsoft.go:111]: Using an access token with scope [User.Read openid profile email] I receive

- from /userinfo

{
    "sub": "AAAAAAAAAAAAAAAAAAAAAENMusLqo_f5vKwPzixWdNS",
    "email": "some@mail.domain",
    "picture": "https://graph.microsoft.com/v1.0/me/photo/$value"
}

- from /me

{
    "@odata.context": "https://graph.microsoft.com/v1.0/$metadata#users/$entity",
    "displayName": "",
    "surname": null,
    "givenName": null,
    "id": "7wt4391vs271486n",
    "userPrincipalName": "some@mail.domain",
    "businessPhones": [],
    "jobTitle": null,
    "mail": null,
    "mobilePhone": null,
    "officeLocation": null,
    "preferredLanguage": null
}

Comment (aeneasr) [reply selfservice/strategy/oidc/provider_microsoft.go:111]: Thank you, this helps!

File: selfservice/strategy/oidc/provider_microsoft.go [hunk 2]
```diff
@@ -67,7 +69,49 @@ func (m *ProviderMicrosoft) Claims(ctx context.Context, exchange *oauth2.Token)
...
+func (m *ProviderMicrosoft) updateSubject(ctx context.Context, claims *Claims, exchange *oauth2.Token) (*Claims, error) {
+	if m.config.SubjectSource == "me" {
+		o, err := m.OAuth2(ctx)
+		if err != nil {
+			return nil, errors.WithStack(herodot.ErrInternalServerError.WithReasonf("%s", err))
+		}
+
+		client := o.Client(ctx, exchange)
+
+		u, err := url.Parse("https://graph.microsoft.com/v1.0/me")
```
Comment (aeneasr) [inline selfservice/strategy/oidc/provider_microsoft.go:89]: Note to self: this is not using the resilient client

File: selfservice/strategy/oidc/provider_microsoft.go [hunk 3]
```diff
@@ -67,7 +69,49 @@ func (m *ProviderMicrosoft) Claims(ctx context.Context, exchange *oauth2.Token)
...
+		}
+
+		resp, err := client.Do(req)
+		if err != nil {
+			return nil, errors.WithStack(herodot.ErrInternalServerError.WithReasonf("%s", err))
+		}
+		defer resp.Body.Close()
+
+		var user struct {
+			Id string `json:"id"`
```
Comment (aeneasr) [inline selfservice/strategy/oidc/provider_microsoft.go:105]:
```suggestion
			ID string `json:"id"`
```

File: selfservice/strategy/oidc/provider_microsoft.go (hunk 1 above)
Comment (aeneasr) [inline selfservice/strategy/oidc/provider_microsoft.go:111]:
```suggestion
		claims.Subject = user.ID
```

**Comment by splaunov** [comment]
I have invited you as a collaborator to the forked repo.

**Comment by aeneasr** [comment]
Thank you! Could you please also update the docs section ( https://www.ory.sh/kratos/docs/next/guides/sign-in-with-github-google-facebook-linkedin#microsoft ) and include how and when to use this feature?

**Comment by aeneasr** [comment]
Ping @splaunov could you please add a section in https://www.ory.sh/docs/kratos/guides/sign-in-with-github-google-facebook-linkedin#microsoft documenting the use of this feature? :)

**Comment by splaunov** [comment]
Yes, sorry for delay. Will do it end of week.

**Comment by aeneasr** [comment]
great, thank you for the update!

**Comment by splaunov** [comment]
Docs PR: https://github.com/ory/docs/pull/706

**Comment by mooijtech** [comment]
Hello,

I am getting the following error when adding subject_source to my OIDC configuration:
```
FATA[2022-04-19T12:07:45+02:00] Unable to instantiate configuration.          audience=application error=map[message:I[#/selfservice/methods/oidc/config/providers/0] S[#/properties/selfservice/properties/methods/properties/oidc/properties/config/properties/providers/items/$ref] doesn't validate with "#/definitions/selfServiceOIDCProvider"
  I[#/selfservice/methods/oidc/config/providers/0] S[#/definitions/selfServiceOIDCProvider/additionalProperties] additionalProperties "subject_source" not allowed] service_name=Ory Kratos service_version=v0.9.0-alpha.3
```
Maybe the definitions have not been updated correctly?
I will try look into this.

**Comment by splaunov** [comment]
> EDIT: Looks like this pull request isn't in v0.9.0-alpha.3

Correct. To test it right now you should build from head of master.

**Pull request closed without merging**